README.md
# Clubhouse Gears

Gears is an extension to Clubhouse.io that enables automating common tasks in a WHEN-THEN pattern. Rules are written in plain Javascript and written directly in Clubhouse. 

<p align="center">
  <a href="http://www.youtube.com/watch?v=SpzzbbauKRA" title="Introduction to Clubhouse Gears">
  <img width="460" height="300" src="assets/youtube.png">
  </a>
</p>

Here's a taste:


- Every third wednesday of the month, create a new story in project "Company maintenance" named "Send invoices". Set due date 7 days later.
- When a story in project "Design" is completed, create a new story in project "Engineering" with the same name and with a relationship to the design story.
- When a story is moved to "Release" column, send an e-mail using Sendgrid.
- For every story created, search Wikipedia for information about name and comment an excerpt. (why would you want this?)

Possibilities really are endless. Also see [Recipes](#Recipes) for code.

# Preparation

1. You need to run Gears somewhere publicly available. For example DigitalOcean, or a home server (with port forwarding). An idea might also be to create a tunnel to your local computer (for local testing).
2. Create a Clubhouse Webhook. Go to clubhouse.io, click your profile picture, Integrations, Webhooks, Add New Webhook. Pick a secret and point it to your publicly available port.

# Installation

**Note of caution:** Code written in Clubhouse stories is executed on the server, meaning Clubhouse members _will be able to run malicious code_. It is recommended to run Gears in an isolated environment such as Docker or a VM on a network that shares no sensitive resources.

## Run locally
```
npm add -g clubhouse-gears

CLUBHOUSE="insert clubhouse API token here" PORT="1337" SECRET="insert clubhouse secret here" gears
```

## Run using Docker

This repository is built at Docker Hub using semver tags. I'd suggest sticking with the major version, unless you're feeling adventurous. See all versions here: https://hub.docker.com/repository/docker/mikabytes/clubhouse-gears/tags

You can also use tag `:latest` for the master branch. Thas is, however, not guaranteed to be stable.

```
sudo docker run -e CLUBHOUSE="insert clubhouse API token here" -e SECRET="insert clubhouse secret here" -p 1337:80 mikabytes/clubhouse-gears:1
```

# How to use

Gears fetches rules from Clubhouse. Any story that has the label "gears" set on it and a name that is formatted like "when (some code goes here)" will be used. See [Recipes](#Recipes) for examples.

When the trigger fires, Gears executes code blocks in the story description (as plain javascript). **Note**: only code blocks are used, so you are free to document your gear as you wish.

# API

When executing code the `action` and `metadata` variables are always available. The `action` object has two origins:
1. From [Webhooks](https://clubhouse.io/api/webhook/v1/) API.
2. From internal Time module. 

`metadata` is all data available, and depends on the nature of the action.

Gears do provide a few convenience variables to the action object, just to ease the strain on your fingers:

| variable | origin | meaning |
| - | - | - |
| id | `action.id` | ID of story, epic, comment, etc. |
| entity_type | `action.entity_type` | "story", "epic", "time" |
| action_type | `action.action`, i.e.  "create", "update" |
| name | `action.name`, the name of the story or the epic in question |
| app_url | `action.appUrl`, the public url |
| changes | `action.changes`, the changes involved in this action, i.e. tags being changed, description updated, etc. |
| author_id | `action.author_id`, ID of the user who created a comment |


## Webhook API

This is prone to change, so not detailed here. See Clubhouse site for information: https://clubhouse.io/api/webhook/v1/

Whenever `references` key is set, Gears will automatically copy the referenced object into the action itself. Again, this is a convenience thing.

## Time API

This is an internal module that submits actions when time has passed various criterias. This is especially suitable when automating the creation of stories that are recurring in nature. Time module always set `entity_type = "time"` and `action = "update"`. The `changes` variable holds one key for each criteria met

| criteria name | possible values | meaning |
| - | - | - |
| minute | 0 to 59 | A new minute |
| hour | 0 to 23 | A new hour |
| day | 1 to 31 | A new calendar day |
| dayOfWeek | 0 to 6 | A new week day, 0 = monday, 6 = sunday |
| week | 1 to 53 | A new calendar week (ISO-8601) |
| daysOfMonth | 1 to 5 | New number of week days in month. I.e. if today is a thursday and daysOfMonth = 3, then today is the third thursday this month |
| month | 1 to 12 | A new month |
| quarter | 1 to 4 | A new quarter |
| year | 1 to 12 | A new year |

These criterias are always present in `metadata` variable. However `action.changes` will only include the ones that are relevant for the current action.

# Extensions

Gears includes two extensions by default:

## `api`

This is an instance of [Clubhouse official API](https://github.com/clubhouse/clubhouse-lib). As it is prone to change, it is not detailed here. 

At the time of this writing they have no documentation either. However, the code is pretty self-explanatory [here](https://github.com/clubhouse/clubhouse-lib/blob/main/src/index.ts)

## `fetch`

This is an implementation of WHATWG fetch for node. Use it as you would normally use `fetch` in the browser.

## Custom

You may add your custom extensions. To do so, create your own NPM package and instantiate Gears this way:

```
npm add mikabytes/clubhouse-gears
```

```
import { addExtension, run } from "clubhouse-gears"

function kaboom(date, warningMessage) {
  robotArm.initiateSelfDestruct(date)
  speakers.tts(warningMessage)
}

addExtension(`kaboom`, kaboom)
run()
```

Your stories can now have code like: `kaboom(new Date().getTime() + 60000, "Self-destruct in T minus 60 seconds")`

# Recipes

Try these examples as a great way to understand how this all comes together.

## Create story for invoicing every first and third wednesday of each month

```
when (changes.dayOfWeek === 2 && [1, 3].includes(changes.daysOfMonth))
```

```javascript
// could also fetch customers from some external API
const customers = [`BigAppleInc`, `DarknessInc`] 
const projects = await api.listProjects()
const deadline = new Date(new Date().getTime() + 60000 * 60 * 24 * 7)
const story = await api.createStory({
  name: `Send invoices`,
  description: `Some desctiption goes here`, `)}`,
  project_id: projects.find((it) => it.name === `Abstrakt`).id,
  deadline: deadline.toISOString(),
})

for (const customer of customers) {
  await createTask(story.id, { description: customer })
}

console.log(`I created ch${story.id}`)
```

## Add information from Wikipedia

This comments on a newly created story with an excerpt from Wikipedia.

```
when (entity_type === `story` && action_type === `create`)
```

```javascript
url = `https://en.wikipedia.org/w/api.php?format=json&action=query&prop=extracts&exsentences=1&titles=${name.split(` `)[0]}`

json = await (fetch(url).then(res => res.json()))

api.createStoryComment(id, Object.values(json.query.pages)[0].extract)
```

## When story is completed, create related story

```
// `new` is true if the `new` state is completed, otherwise it is false. There is also `old`.
when (changes.completed && changes.completed.new)
```

```javascript
const whenFinishedProjectName = `design`
const thenCreateProjectName = `engineering`

const projects = await api.listProjects()
const project_id = projects.find(it => it.name === whenFinishedProjectName).id
const story = await api.getStory(id)

if (story.project_id === project_id) {
  const newStory = await api.createStory({ 
    name: story.name,
    description: `See linked story for design.`,
    project_id: projects.find(it => it.name === thenCreateProjectName).id,
    epic_id: story.epic_id,
  })

  await api.createStoryLink({
    object_id: story.id,
    subject_id: newStory.id,
    verb: `relates to`, // can also be 'blocks' or 'duplicates'
  })
}

console.log(`I just created this story ch${newStory.id}`)
```

## Add a command to clear all comments on a story

When anyone writes a comment with text `/clear`, all comments is deleted on that story. I find this especially useful when clearing error messages in a Gears story.

```
when (entity_type === `story-comment` && action_type === `create`)
```

```javascript
if (action.text === `/clear`) {
  storyId = action.app_url.match(/story\/(\d+)/)[1]
  story = await api.getStory(storyId)
  for (const comment of story.comments) {
    try {
      await api.deleteStoryComment(storyId, comment.id)
    } catch(e) {
      // api always throws error here, even if the call was successful. There's a request to fix that: https://github.com/clubhouse/clubhouse-lib/pull/98
    }
  }
}
```

# Roadmap

- Custom external triggers: When, or if, there's interest, I'll add a feature to provide additional triggers. This would enable external integration in some ways that are currently limited. It can still be done with some form of polling (based on the `minutes` event from `time` module).


CHANGELOG.md
# 1.0.2

Added support for automatic Docker builds, and added documentation

# 1.0.1

Clubhouse started to deactivate the webhook if it didn't get a 200 response code. This has now been fixed.


storiesToRules.test.js
import assert from "assert"
import _storiesToRules from "./storiesToRules.js"

// convenience function for the stories as values are already known, not await'd to be fetched over network
async function storiesToRules(stories) {
  let all = []

  for await (const story of _storiesToRules(stories)) {
    all.push(story)
  }

  return all
}

describe(`storiesToRules`, async () => {
  it(`should not include stories whose title doesn't start with 'when'`, async () => {
    const rules = await storiesToRules([
      { name: `Not starting with 'when'`, description: `` },
    ])

    assert.equal(rules.length, 0)
  })

  it(`should include stories that starts with 'when'`, async () => {
    const rules = await storiesToRules([
      { name: `when()`, description: `` },
      { name: `wHEN()`, description: `` },
      { name: `when ()`, description: `` },
      { name: `wHEN ()`, description: `` },
    ])

    assert.equal(rules.length, 4)
  })

  it(`should include story as third argument`, async () => {
    const rules = await storiesToRules([
      { id: 3, name: `when()`, description: `` },
    ])

    assert.equal(rules[0][2].id, 3)
  })

  it(`should return an array with condition as first entry`, async () => {
    const rules = await storiesToRules([
      { name: `when (some javascript here)`, description: `` },
    ])

    assert.equal(rules[0][0], `some javascript here`)
  })

  it(`should return an array with any markdown code blocks as WHEN`, async () => {
    const rules = await storiesToRules([
      {
        name: `when (condition)`,
        description:
          "Not here\n```javascript\nthis is\ncode\n\nmultiline\n```\nbut this is not included\n```though this is``` and again not",
      },
    ])

    assert.equal(rules[0][1], `this is\ncode\n\nmultiline\n\nthough this is`)
  })
})


getTimeBasedChanges.test.js
import assert from "assert"
import getTimeBasedChanges from "./getTimeBasedChanges.js"

describe(`getTimeBasedChanges`, () => {
  it(`should always include minute as change, and rest in reference`, () => {
    const [changes, reference] = getTimeBasedChanges(
      new Date(2021, 2, 11, 17, 20)
    )
    assert.equal(changes.minute, 20)
    assert.equal(reference.minute, 20)
    assert.equal(reference.hour, 17)
    assert.equal(reference.dayOfWeek, 3)
    assert.equal(reference.day, 11)
    assert.equal(reference.week, 10)
    assert.equal(reference.daysOfMonth, 2) // 2nd thursday of month
    assert.equal(reference.month, 2)
    assert.equal(reference.year, 2021)
  })

  it(`should include new hour`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 0, 1, 10, 0))
    assert.deepEqual(changes, { minute: 0, hour: 10 })
  })

  it(`should include new day, dayOfWeek`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 2, 11, 0, 0))
    assert.deepEqual(changes, { minute: 0, hour: 0, day: 11, dayOfWeek: 3 })
  })

  it(`should include new week`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 2, 15, 0, 0))
    assert.deepEqual(changes, {
      minute: 0,
      hour: 0,
      day: 15,
      dayOfWeek: 0,
      week: 11,
    })
  })

  it(`should include new month`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 2, 1, 0, 0))
    assert.deepEqual(changes, {
      minute: 0,
      hour: 0,
      day: 1,
      dayOfWeek: 0,
      week: 9,
      month: 2,
    })
  })

  it(`should include new year`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 0, 1, 0, 0))
    assert.deepEqual(changes, {
      minute: 0,
      hour: 0,
      day: 1,
      dayOfWeek: 4,
      month: 0,
      year: 2021,
    })
  })
})


src/getTimeBasedChanges.test.js
import assert from "assert"
import getTimeBasedChanges from "./getTimeBasedChanges.js"

describe(`getTimeBasedChanges`, () => {
  it(`should always include minute as change, and rest in reference`, () => {
    const [changes, reference] = getTimeBasedChanges(
      new Date(2021, 2, 11, 17, 20) // March 11, 5.20pm, 2021
    )
    assert.equal(changes.minute, 20)
    assert.equal(reference.minute, 20)
    assert.equal(reference.hour, 17)
    assert.equal(reference.dayOfWeek, 3)
    assert.equal(reference.day, 11)
    assert.equal(reference.week, 10)
    assert.equal(reference.daysOfMonth, 2) // 2nd thursday of month
    assert.equal(reference.month, 2)
    assert.equal(reference.year, 2021)
  })

  it(`should include new hour`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 0, 1, 10, 0))
    assert.deepEqual(changes, { minute: 0, hour: 10 })
  })

  it(`should include new day, dayOfWeek`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 2, 11, 0, 0))
    assert.deepEqual(changes, { minute: 0, hour: 0, day: 11, dayOfWeek: 3 })
  })

  it(`should include new week`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 2, 15, 0, 0))
    assert.deepEqual(changes, {
      minute: 0,
      hour: 0,
      day: 15,
      dayOfWeek: 0,
      week: 11,
    })
  })

  it(`should include Q1`, () => {
    let [changes] = getTimeBasedChanges(new Date(2021, 0, 1, 0, 0))
    assert.equal(changes.quarter, 1)
  })

  it(`should include Q2`, () => {
    let [changes] = getTimeBasedChanges(new Date(2021, 3, 1, 0, 0))
    assert.equal(changes.quarter, 2)
  })
  it(`should include Q3`, () => {
    let [changes] = getTimeBasedChanges(new Date(2021, 6, 1, 0, 0))
    assert.equal(changes.quarter, 3)
  })
  it(`should include Q4`, () => {
    let [changes] = getTimeBasedChanges(new Date(2021, 9, 1, 0, 0))
    assert.equal(changes.quarter, 4)
  })

  it(`should include new month`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 2, 1, 0, 0))
    assert.deepEqual(changes, {
      minute: 0,
      hour: 0,
      day: 1,
      dayOfWeek: 0,
      week: 9,
      month: 2,
    })
  })

  it(`should include new year`, () => {
    const [changes] = getTimeBasedChanges(new Date(2021, 0, 1, 0, 0))
    assert.deepEqual(changes, {
      minute: 0,
      hour: 0,
      day: 1,
      dayOfWeek: 4,
      month: 0,
      quarter: 1,
      year: 2021,
    })
  })
})


src/getTimeBasedChanges.js
export default function getTimeBasedChanges(date) {
  const reference = {
    minute: date.getMinutes(),
    hour: date.getHours(),
    dayOfWeek: (7 + date.getDay() - 1) % 7, // day of week, starting monday,
    day: date.getDate(), // note: this is the only one that's not zero-based
    week: getWeekNumber(date),
    daysOfMonth: getDaysOfMonth(date),
    month: date.getMonth(),
    quarter: getQuarter(date),
    year: date.getFullYear(),
  }

  const changes = {
    minute: reference.minute,
  }

  if (reference.minute === 0) {
    changes.hour = reference.hour

    if (reference.hour === 0) {
      changes.day = reference.day
      changes.dayOfWeek = reference.dayOfWeek

      if (reference.dayOfWeek === 0) {
        changes.week = reference.week
      }

      if (reference.day === 1) {
        changes.month = reference.month

        if ([0, 3, 6, 9].includes(reference.month)) {
          changes.quarter = reference.quarter
        }

        if (reference.month === 0) {
          changes.year = reference.year
        }
      }
    }
  }

  return [changes, reference]
}

// from https://stackoverflow.com/questions/6117814/get-week-of-year-in-javascript-like-in-php
function getWeekNumber(d) {
  // Copy date so don't modify original
  d = new Date(Date.UTC(d.getFullYear(), d.getMonth(), d.getDate()))
  // Set to nearest Thursday: current date + 4 - current day number
  // Make Sunday's day number 7
  d.setUTCDate(d.getUTCDate() + 4 - (d.getUTCDay() || 7))
  // Get first day of year
  var yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1))
  // Calculate full weeks to nearest Thursday
  var weekNo = Math.ceil(((d - yearStart) / 86400000 + 1) / 7)
  return weekNo
}

function getDaysOfMonth(d) {
  let count = 0
  const month = d.getMonth()

  while (d.getMonth() === month) {
    count += 1
    d = new Date(d.getTime() - 7 * 24 * 60 * 60 * 1000)
  }

  return count
}

function getQuarter(d) {
  let m = Math.floor(d.getMonth() / 3) + 1
  return m
}


src/index.js
import express from "express"
import crypto from "crypto"
import bodyParser from "body-parser"
import getTimeBasedChanges from "./getTimeBasedChanges.js"
import morgan from "morgan"
import Clubhouse from "clubhouse-lib"
import fetch from "node-fetch"
import storiesToRules from "./storiesToRules.js"
import util from "util"

const extensions = []

export function addExtension(name, implementation) {
  extensions.push([name, implementation])
}

const api = Clubhouse.create(process.env.CLUBHOUSE)

addExtension(`fetch`, fetch)
addExtension(`api`, api)

const port = process.env.PORT || 80
const defaultSecret = `oh-so-secret`
const secret = process.env.SECRET || defaultSecret

let rules = new Map()

async function setupRules() {
  rules.clear()

  const localVariables = `
    const id = action.id
    const entity_type = action.entity_type
    const action_type = action.action
    const story_type = action.story_type
    const name = action.name
    const app_url = action.appUrl
    const changes = action.changes || {}
    const author_id = action.author_id
  `

  for await (const [whenCode, thenCode, story] of storiesToRules(
    downloadStories()
  )) {
    const c = createConsole({ story })

    try {
      const whenFunc = new Function(
        `action`,
        `metadata`,
        `console`,
        ...extensions.map(([it]) => it),
        `
          ${localVariables}

          return (${whenCode})
        `
      )
      whenFunc.story = story
      whenFunc.code = whenCode

      const thenFunc = new Function(
        `action`,
        `metadata`,
        `console`,
        ...extensions.map(([it]) => it),
        `
          ${localVariables}

          return (async function(){ ${thenCode} })()
        `
      )
      thenFunc.story = story
      thenFunc.code = thenCode

      when(whenFunc).then(thenFunc)
    } catch (e) {
      c.error(e)
    }
  }

  // add our own hook to update scripts whenever a matching gears rule is updated
  when(
    (action) =>
      action.entity_type === `story` &&
      action.name.match(/^when ?\([^)]+\)$/i) &&
      (!action.changes ||
        action.changes.name ||
        action.changes.description ||
        action.changes.label_ids)
  ).then(({ name }) => {
    setupRules()
    console.log(`=== Updated rules because a change was detected to: ${name}`)
  })
}

async function* downloadStories() {
  let res = await api.searchStories(`!is:archived label:"gears"`)

  yield* res.data.values()

  while (res.next) {
    res = await res.fetchNext()
    yield* res.data.values()
  }
}

const app = express()
app.use(morgan(`dev`))
app.use(bodyParser.text({ type: `application/json` }))

if (secret === defaultSecret) {
  console.warn(
    `Secret isn't configured, so we are not protected from malicious otters.`
  )
}

app.post(`/`, async (req, res) => {
  const signature = req.get(`Clubhouse-Signature`)

  if (!signature || !correctDigest(req.body, signature)) {
    console.log(`Invalid signature ${signature}`)
    return res.status(401).json({ message: `You are unauthorized.` })
  } else {
    res.send(200)
  }

  let json
  try {
    json = JSON.parse(req.body)
  } catch (e) {
    console.error(`Could not parse:\n${req.body}`)
    return res.status(400).json({ message: `You sent an invalid body.` })
  }

  if (!json.references) {
    json.references = []
  }

  for (const action of json.actions) {
    if (action && action.changes) {
      let changes = action.changes
      for (const key in changes) {
        if (
          changes[key] &&
          changes[key].new !== undefined &&
          changes[key].old !== undefined
        ) {
          // this is a reference, automatically copy it from references
          const newRef = json.references.find(
            (it) => it.id === changes[key].new
          )
          const oldRef = json.references.find(
            (it) => it.id === changes[key].old
          )

          if (newRef && oldRef) {
            changes[key].new = newRef
            changes[key].old = oldRef
          }
        }
      }
    }
  }

  console.log(JSON.stringify(json, null, 4))

  await broadcast(json.actions, json)
})

function createConsole(source) {
  const id = source.story && source.story.id

  const c = {
    async log(...stuff) {
      console.log(source.code, source.action, ...stuff)
      if (id) {
        await api.createStoryComment(
          id,
          stuff
            .map((it) =>
              typeof it === `string` ? it : "```\n" + stringify(it) + "\n```"
            )
            .join(`\n\n`)
        )
      }
    },
    async error(...stuff) {
      console.error(source.code, source.action, ...stuff)

      if (id) {
        await api.createStoryComment(
          id,
          `:bangbang:` +
            stuff
              .map((it) =>
                typeof it === `string` ? it : "```\n" + stringify(it) + "\n```"
              )
              .join(`\n\n`)
        )
      }
    },
  }
  c.info = c.log
  return c
}

async function broadcast(actions, metadata) {
  for (const [when, thens] of rules) {
    for (const action of actions) {
      let pass
      const c = createConsole(when)
      try {
        pass = when(action, metadata, c, ...extensions.map(([_, it]) => it))
      } catch (e) {
        c.error(e)
        continue
      }

      if (pass) {
        for (const then of thens) {
          const c = createConsole(then)
          try {
            await then(action, metadata, c, ...extensions.map(([_, it]) => it))
          } catch (e) {
            c.error(e)
          }
        }
      }
    }
  }
}

function correctDigest(body, signature) {
  const hmac = crypto.createHmac(`sha256`, secret, { encoding: `utf8` })
  hmac.update(body)
  const expectedSignature = hmac.digest(`hex`)

  return expectedSignature === signature
}

export function when(predicate) {
  if (!rules.has(predicate)) {
    rules.set(predicate, new Set())
  }
  return {
    then(action) {
      rules.get(predicate).add(action)
    },
  }
}

function broadcastTimeBasedActions() {
  const date = new Date()
  const [changes, reference] = getTimeBasedChanges(date)
  broadcast(
    [{ id: date.getTime(), entity_type: `time`, action: `update`, changes }],
    reference
  )
}

export async function run() {
  try {
    setInterval(broadcastTimeBasedActions, 60000)

    app.listen(port)

    await setupRules()
  } catch (e) {
    console.error(e)
  }
}

function stringify(obj) {
  return util.inspect(obj)
}
